##### Foundations of Machine Learning

#### Letter Challenge Unlabelled

In [67]:
%cd D:\

D:\


In [68]:
import pandas as pd

In [69]:
df=pd.read_csv('letter-challenge-unlabeled.csv',sep=",")  #Loading Dataset

In [70]:
df.head()

x-box  y-box  width  height  onpixels  x-bar  y-bar  x2bar  y2bar  xybar  \
0      6      9      9       7         6      8      8      4      1      7   
1      2      9      3       7         1     12      2     10      4     13   
2      5     10      5       5         4      7      7      4      6      8   
3      1      3      2       1         1      8      7      5      3      9   
4      1      0      2       1         1      8      7      6      4      6   

   x2ybr  xy2br  x-edge  xedgevy  y-edge  yedgevx letter  
0      9      8       7       11       0        8      -  
1      6     13       1        6       0        8      -  
2      5      7       5       10       6        6      +  
3      6      8       2        8       2        8      +  
4      6      9       2        8       3        8      +

In [71]:
df['letter'].value_counts()

?    10000
-     7240
+     2760
Name: letter, dtype: int64

In [72]:
df.shape

(20000, 17)

In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
x-box       20000 non-null int64
y-box       20000 non-null int64
width       20000 non-null int64
height      20000 non-null int64
onpixels    20000 non-null int64
x-bar       20000 non-null int64
y-bar       20000 non-null int64
x2bar       20000 non-null int64
y2bar       20000 non-null int64
xybar       20000 non-null int64
x2ybr       20000 non-null int64
xy2br       20000 non-null int64
x-edge      20000 non-null int64
xedgevy     20000 non-null int64
y-edge      20000 non-null int64
yedgevx     20000 non-null int64
letter      20000 non-null object
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


In [74]:
df.describe()

x-box         y-box         width       height      onpixels  \
count  20000.000000  20000.000000  20000.000000  20000.00000  20000.000000   
mean       4.023550      7.035500      5.121850      5.37245      3.505850   
std        1.913212      3.304555      2.014573      2.26139      2.190458   
min        0.000000      0.000000      0.000000      0.00000      0.000000   
25%        3.000000      5.000000      4.000000      4.00000      2.000000   
50%        4.000000      7.000000      5.000000      6.00000      3.000000   
75%        5.000000      9.000000      6.000000      7.00000      5.000000   
max       15.000000     15.000000     15.000000     15.00000     15.000000   

              x-bar         y-bar         x2bar         y2bar         xybar  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.897600      7.500450      4.628600      5.178650      8.282050   
std        2.026035      2.325354      2.699968      2.380823      2.488475   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.000000      6.000000      3.000000      4.000000      7.000000   
50%        7.000000      7.000000      4.000000      5.000000      8.000000   
75%        8.000000      9.000000      6.000000      7.000000     10.000000   
max       15.000000     15.000000     15.000000     15.000000     15.000000   

             x2ybr         xy2br        x-edge       xedgevy        y-edge  \
count  20000.00000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.45400      7.929000      3.046100      8.338850      3.691750   
std        2.63107      2.080619      2.332541      1.546722      2.567073   
min        0.00000      0.000000      0.000000      0.000000      0.000000   
25%        5.00000      7.000000      1.000000      8.000000      2.000000   
50%        6.00000      8.000000      3.000000      8.000000      3.000000   
75%        8.00000      9.000000      4.000000      9.000000      5.000000   
max       15.00000     15.000000     15.000000     15.000000     15.000000   

           yedgevx  
count  20000.00000  
mean       7.80120  
std        1.61747  
min        0.00000  
25%        7.00000  
50%        8.00000  
75%        9.00000  
max       15.00000

In [75]:
df.isnull().sum()   #Here no missing values but letter has '?' which is necessarily to be -/+

x-box       0
y-box       0
width       0
height      0
onpixels    0
x-bar       0
y-bar       0
x2bar       0
y2bar       0
xybar       0
x2ybr       0
xy2br       0
x-edge      0
xedgevy     0
y-edge      0
yedgevx     0
letter      0
dtype: int64

In [76]:
df['letter']=df['letter'].str.replace('?','')  #Replacing ? with blanks

In [77]:
import numpy as np
df['letter']=df['letter'].replace(r'\s+( +\.)|#',np.nan,regex=True).replace('',np.nan) #Replaces blanks with NaN

In [78]:
df['letter'].value_counts()

-    7240
+    2760
Name: letter, dtype: int64

In [79]:
df.isnull().sum()      # in Letter 10000 Missing Values are there

x-box           0
y-box           0
width           0
height          0
onpixels        0
x-bar           0
y-bar           0
x2bar           0
y2bar           0
xybar           0
x2ybr           0
xy2br           0
x-edge          0
xedgevy         0
y-edge          0
yedgevx         0
letter      10000
dtype: int64

In [80]:
df['letter']=df['letter'].fillna('-')     # Filling Blanks with '-'

In [81]:
df['letter'].value_counts() # Blanks are replaced with '-'

-    17240
+     2760
Name: letter, dtype: int64

#### Variance of Numerical Columns

In [82]:
np.var(df)

x-box        3.660195
y-box       10.919540
width        4.058303
height       5.113631
onpixels     4.797866
x-bar        4.104614
y-bar        5.407000
x2bar        7.289462
y2bar        5.668034
xybar        6.192198
x2ybr        6.922184
xy2br        4.328759
x-edge       5.440475
xedgevy      2.392231
y-edge       6.589532
yedgevx      2.616079
dtype: float64

In [83]:
df.columns

Index(['x-box', 'y-box', 'width', 'height', 'onpixels', 'x-bar', 'y-bar',
       'x2bar', 'y2bar', 'xybar', 'x2ybr', 'xy2br', 'x-edge', 'xedgevy',
       'y-edge', 'yedgevx', 'letter'],
      dtype='object')

In [84]:
data=df[['x-box', 'y-box', 'width', 'height', 'onpixels', 'x-bar', 'y-bar',
       'x2bar', 'y2bar', 'xybar', 'x2ybr', 'xy2br', 'x-edge', 'xedgevy',
       'y-edge', 'yedgevx']]

### Normalization 

In [85]:
def normalize(data):
    result = data.copy()
    for feature_name in data.columns:
        max_value = data[feature_name].max()
        min_value = data[feature_name].min()
        result[feature_name] = (data[feature_name] - min_value) / (max_value - min_value)
    return result

In [86]:
datan=normalize(data)

In [87]:
datan.head()

x-box     y-box     width    height  onpixels     x-bar     y-bar  \
0  0.400000  0.600000  0.600000  0.466667  0.400000  0.533333  0.533333   
1  0.133333  0.600000  0.200000  0.466667  0.066667  0.800000  0.133333   
2  0.333333  0.666667  0.333333  0.333333  0.266667  0.466667  0.466667   
3  0.066667  0.200000  0.133333  0.066667  0.066667  0.533333  0.466667   
4  0.066667  0.000000  0.133333  0.066667  0.066667  0.533333  0.466667   

      x2bar     y2bar     xybar     x2ybr     xy2br    x-edge   xedgevy  \
0  0.266667  0.066667  0.466667  0.600000  0.533333  0.466667  0.733333   
1  0.666667  0.266667  0.866667  0.400000  0.866667  0.066667  0.400000   
2  0.266667  0.400000  0.533333  0.333333  0.466667  0.333333  0.666667   
3  0.333333  0.200000  0.600000  0.400000  0.533333  0.133333  0.533333   
4  0.400000  0.266667  0.400000  0.400000  0.600000  0.133333  0.533333   

     y-edge   yedgevx  
0  0.000000  0.533333  
1  0.000000  0.533333  
2  0.400000  0.400000  
3  0.133333  0.533333  
4  0.200000  0.533333

In [88]:
df['letter']=df['letter'].str.replace('+','0')     #Replacing '+' with 0
df['letter']=df['letter'].str.replace('-','1')     #Replacing '-' with 1  

In [89]:
df['letter'].value_counts()

1    17240
0     2760
Name: letter, dtype: int64

In [90]:
df.isnull().sum()    #No Missing Values in Dataset

x-box       0
y-box       0
width       0
height      0
onpixels    0
x-bar       0
y-bar       0
x2bar       0
y2bar       0
xybar       0
x2ybr       0
xy2br       0
x-edge      0
xedgevy     0
y-edge      0
yedgevx     0
letter      0
dtype: int64

In [91]:
df.info()   #Lets convert this letter from object datatype to int

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
x-box       20000 non-null int64
y-box       20000 non-null int64
width       20000 non-null int64
height      20000 non-null int64
onpixels    20000 non-null int64
x-bar       20000 non-null int64
y-bar       20000 non-null int64
x2bar       20000 non-null int64
y2bar       20000 non-null int64
xybar       20000 non-null int64
x2ybr       20000 non-null int64
xy2br       20000 non-null int64
x-edge      20000 non-null int64
xedgevy     20000 non-null int64
y-edge      20000 non-null int64
yedgevx     20000 non-null int64
letter      20000 non-null object
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


In [92]:
df['letter']=df['letter'].astype('int')

In [93]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
x-box       20000 non-null int64
y-box       20000 non-null int64
width       20000 non-null int64
height      20000 non-null int64
onpixels    20000 non-null int64
x-bar       20000 non-null int64
y-bar       20000 non-null int64
x2bar       20000 non-null int64
y2bar       20000 non-null int64
xybar       20000 non-null int64
x2ybr       20000 non-null int64
xy2br       20000 non-null int64
x-edge      20000 non-null int64
xedgevy     20000 non-null int64
y-edge      20000 non-null int64
yedgevx     20000 non-null int64
letter      20000 non-null int32
dtypes: int32(1), int64(16)
memory usage: 2.5 MB


#### Logistic Regression

#### Our Features are in first 16 Columns so saving all these features in x
#### Target Variable is in 17 column(Letter) and saving in y

In [94]:
x = datan.iloc[:,[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]].values
y = df.iloc[:,16].values

#### Splitting the Dataset 

In [95]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.01,random_state=0)

In [96]:
from sklearn.preprocessing import StandardScaler
sc_x= StandardScaler()
x_train=sc_x.fit_transform(x_train)
x_test=sc_x.transform(x_test)

#### Importing Logistic Regression Library

In [97]:
from sklearn.linear_model import LogisticRegression
classifier=LogisticRegression(random_state=0)
classifier.fit(x_train,y_train)

C:\Users\nsidd\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=0, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

In [98]:
x_test

array([[ 1.03338219,  0.89760457,  1.92441285, ...,  1.72126241,
        -1.43737405,  0.12322625],
       [ 1.55609238,  1.20020397,  0.9320341 , ...,  2.36804653,
        -0.65856535, -2.35004272],
       [-0.53474836,  0.59500517, -0.55653402, ..., -0.21908995,
         1.67786074,  0.12322625],
       ...,
       [ 0.51067201,  0.89760457,  1.42822348, ...,  0.42769417,
        -1.0479697 , -0.49509099],
       [ 2.07880256,  0.59500517,  0.43584473, ...,  0.42769417,
         0.12024334,  0.12322625],
       [-1.58016872, -1.52319063, -1.54891277, ...,  0.42769417,
        -0.26916101,  0.7415435 ]])

In [99]:
y_pred=classifier.predict(x_test)

In [100]:
y_pred

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1])

In [101]:
predicted_class=classifier.predict(x_test)

#### Accuracy Score of Logistic Regression

In [102]:
from sklearn.metrics import accuracy_score

In [103]:
accuracy = accuracy_score(y_test,predicted_class)
parameters = classifier.coef_
print(accuracy)

0.885


#### Accuracy Score of Logistic Regression is 88.5%

#### Confusion Matrix

In [104]:
from sklearn.metrics import confusion_matrix 

In [105]:
cm=confusion_matrix(y_test,y_pred)

In [143]:
cm #23 observations were predicted wrong

array([[  1,  20],
       [  3, 176]], dtype=int64)

#### Neural Network

In [107]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.01,random_state=0)

In [125]:
x = datan.iloc[:,[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]].values
y = df.iloc[:,16].values

#### Importing MultiLayerPerceptron from Neural Network

In [126]:
from sklearn.neural_network import MLPClassifier

In [127]:
nnmodel = MLPClassifier(hidden_layer_sizes=(100,40),random_state=0)

In [128]:
nnmodel.fit(x_train,y_train)

C:\Users\nsidd\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(100, 40), learning_rate='constant',
              learning_rate_init=0.001, max_iter=200, momentum=0.9,
              n_iter_no_change=10, nesterovs_momentum=True, power_t=0.5,
              random_state=0, shuffle=True, solver='adam', tol=0.0001,
              validation_fraction=0.1, verbose=False, warm_start=False)

In [129]:
nnpredicted_class=nnmodel.predict(x_test)

In [130]:
accuracy_score(y_test,nnpredicted_class)

0.89

In [113]:
from sklearn.metrics import classification_report
print(classification_report(y_test,nnpredicted_class))  #Classification Report

              precision    recall  f1-score   support

           0       0.48      0.62      0.54        21
           1       0.95      0.92      0.94       179

    accuracy                           0.89       200
   macro avg       0.72      0.77      0.74       200
weighted avg       0.90      0.89      0.90       200



In [133]:
from sklearn.metrics import confusion_matrix
nn_cm=confusion_matrix(y_test, nnpredicted_class)


In [151]:
nn_cm     #Out of 200 obs 178 were predictions were True, 22 were Falsely predicted

array([[ 13,   8],
       [ 14, 165]], dtype=int64)

#### Random Forest

In [139]:
#from sklearn.linear_model import LogisticRegression 
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score 

#model = LogisticRegression()
model = RandomForestClassifier(n_estimators=100)

model.fit(x, y)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [140]:
predicted_classes = model.predict(x)

In [141]:
accuracy = accuracy_score(y.flatten(),predicted_classes)
print(accuracy)

0.995


#### Accuracy Score of Random Forest is 99.5%

#### There we go after executing Random Forest we got the accuracy of 99.5%

#### So,From all the above Algorithms Random Forest has the Best Accuracy to predict the Target Variable